src/py/flwr/server/strategy/qfedavg.py

```diff
@@ -191,8 +191,29 @@ def norm_grad(grad_list: NDArrays) -> float:
         eval_result = self.evaluate(
             server_round, ndarrays_to_parameters(weights_before)
         )
+        # Ensure `loss` is always defined
+        # Aggregate fit results
+        losses = [
+            fit_res.loss
+            for _, fit_res in results
+            if hasattr(fit_res, "loss") and not np.isnan(fit_res.loss)
+        ]
+
         if eval_result is not None:
             loss, _ = eval_result
+        elif losses:
+            loss = np.mean(losses)
+            log(
+                WARNING,
+                "Using dynamic fallback loss based on historical data: %s",
+                loss,
+            )
```

Comment:
Will never be evaluated (linked to previous comment)

Reply:
removed elif statement, the other cases handles it.